pkg/network/cache/cache.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package cache
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	kfs "kubevirt.io/kubevirt/pkg/os/fs"
+)
+
+type Cache struct {
+	path string
+	fs   cacheFS
+}
+
+type cacheFS interface {
+	Stat(name string) (os.FileInfo, error)
+	MkdirAll(path string, perm os.FileMode) error
+	RemoveAll(path string) error
+	ReadFile(filename string) ([]byte, error)
+	WriteFile(filename string, data []byte, perm fs.FileMode) error
+}
+
+type CacheCreator struct{}
+
+func (_ CacheCreator) New(filePath string) *Cache {
+	return NewCustomCache(filePath, kfs.New())
+}
+
+func NewCustomCache(path string, fs cacheFS) *Cache {
+	return &Cache{path, fs}
+}
+
+func (c Cache) Entry(path string) (Cache, error) {
+	fileInfo, err := c.fs.Stat(c.path)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return Cache{}, fmt.Errorf("unable to define entry: %v", err)
+		}
+	} else if !fileInfo.IsDir() {
+		return Cache{}, fmt.Errorf("unable to define entry: parent cache has an existing store")
+	}
+
+	return Cache{
+		path: filepath.Join(c.path, path),
```

Comment:
The cache backend is a folder or file.
If it is a folder, it means it has "childrens", and it may have multiple such "childrens".
But the cache one wants to directly write or read from, must be one with a backed file.

Once you write to the cache, it creates a file in the path.
So this is valid:
```golang
rootCache := cache.New("/cache")
myCache := rootCache.Entry("mycache")
yourCache := rootCache.Entry("yourcache")

myCache.Write(...)    // Backend file is created here
yourCache.Write(...)  // Backend file is created here

rootCache.Write(...) // this will fail, as it is not a leaf cache.
```